### DISCUSSION TITLE
nginx with internal docker network

### DISCUSSION THREAD

**Discussion by dannongruver** [open-webui/open-webui#1203] [General]
I'm attempting to run open-webui in docker on an **internal** network. Since the network's internal it can't access the host/internet, I'm also attempting to reverse proxy to the host via nginx. I'm successfully able to serve up static pages from open-webui (eg favicon.png) using nginx but not the dynamic pages. Also open-webui works without using this internal network and reverse proxy.

Although this is a docker/nginx question, the reason I'm asking this here is because it appears that open-webui is running on Uvicorn. According to Uvicorn's documentation (https://www.uvicorn.org/deployment/#running-behind-nginx) these settings may need to be set: `--forwarded-allow-ips='*'` and `--proxy-headers`. 

My questions: 
1) Is there a way to expose those Uvicorn properties? 
2) If this works, can the correct yaml for this scenario be added to docs.

_I'm not an expert at any of this so I'm sure I have the wrong configuration somewhere._ 

Here's my docker-compose yaml: 
```
version: '3.8'
services:
    ollama:
        deploy:
            resources:
                reservations:
                    devices:
                        - driver: nvidia
                          count: all
                          capabilities:
                              - gpu
        volumes:
            - C:\Projects\LocalLLM\ollama:/root/.ollama
        ports:
            - 11434:11434
        container_name: ollama
        image: ollama/ollama
        command: 
        networks:
          isolated:
    open-webui:
        ports:
            - 8080:8080
        extra_hosts:
            - host.docker.internal:host-gateway
        container_name: open-webui
        restart: always
        image: ghcr.io/open-webui/open-webui:main
        depends_on:
            - ollama
        networks:
            - isolated

    nginx:
        container_name: nginx
        ports:
            - 80:80
        image: nginx:latest
        depends_on:
            - open-webui
        networks:
            - isolated    
            - internet
networks:
    isolated:
        driver: bridge
        internal: true
    internet:
        driver: bridge
```

Here's my nginx config (copied from https://www.uvicorn.org/deployment/#running-behind-nginx):
```
server {
    listen 80;
    client_max_body_size 4G;

    server_name localhost;

    location / {
      proxy_set_header Host $http_host;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection $connection_upgrade;
      proxy_redirect off;
      proxy_buffering off;
      proxy_pass http://open-webui:8080;
    }

    location /static {
      # path for static files
      root /path/to/app/static;
    }
  }

  map $http_upgrade $connection_upgrade {
    default upgrade;
    '' close;
  }

  upstream uvicorn {
    server unix:/tmp/uvicorn.sock;
  }

```
here are open-webui container logs
```
2024-03-18 18:35:11 No WEBUI_SECRET_KEY provided
2024-03-18 18:35:11 Generating WEBUI_SECRET_KEY
2024-03-18 18:35:11 Loading WEBUI_SECRET_KEY from .webui_secret_key
2024-03-18 18:35:19 
2024-03-18 18:35:19 #------------------------------------------------------------#
2024-03-18 18:35:19 #                                                            #
2024-03-18 18:35:19 #           'It would help me if you could add...'            #
2024-03-18 18:35:19 #        https://github.com/BerriAI/litellm/issues/new        #
2024-03-18 18:35:19 #                                                            #
2024-03-18 18:35:19 #------------------------------------------------------------#
2024-03-18 18:35:19 
2024-03-18 18:35:19  Thank you for using LiteLLM! - Krrish & Ishaan
2024-03-18 18:35:19 
2024-03-18 18:35:19 
2024-03-18 18:35:19 
2024-03-18 18:35:19 Give Feedback / Get Help: https://github.com/BerriAI/litellm/issues/new
2024-03-18 18:35:19 
2024-03-18 18:35:19 
2024-03-18 18:35:19 INFO:     192.168.128.1:0 - "GET /favicon.png HTTP/1.0" 200 OK
2024-03-18 18:35:26 INFO:     192.168.128.1:0 - "GET / HTTP/1.0" 304 Not Modified
2024-03-18 18:35:15 INFO:     Started server process [1]
2024-03-18 18:35:15 INFO:     Waiting for application startup.
2024-03-18 18:35:15 INFO:     Application startup complete.
2024-03-18 18:35:15 INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
```

here are nginx container logs:```
```
2024-03-18 18:35:12 /docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
2024-03-18 18:35:12 /docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
2024-03-18 18:35:12 /docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
2024-03-18 18:35:12 10-listen-on-ipv6-by-default.sh: info: Getting the checksum of /etc/nginx/conf.d/default.conf
2024-03-18 18:35:12 10-listen-on-ipv6-by-default.sh: info: /etc/nginx/conf.d/default.conf differs from the packaged version
2024-03-18 18:35:12 /docker-entrypoint.sh: Sourcing /docker-entrypoint.d/15-local-resolvers.envsh
2024-03-18 18:35:12 /docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
2024-03-18 18:35:12 /docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
2024-03-18 18:35:12 /docker-entrypoint.sh: Configuration complete; ready for start up
2024-03-18 18:35:19 192.168.128.1 - - [18/Mar/2024:23:35:19 +0000] "GET /favicon.png HTTP/1.1" 200 6161 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36 Edg/122.0.0.0" "-"
2024-03-18 18:35:26 192.168.128.1 - - [18/Mar/2024:23:35:26 +0000] "GET / HTTP/1.1" 304 0 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36 Edg/122.0.0.0" "-"
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: using the "epoll" event method
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: nginx/1.25.4
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: built by gcc 12.2.0 (Debian 12.2.0-14) 
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: OS: Linux 5.15.146.1-microsoft-standard-WSL2
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker processes
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 29
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 30
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 31
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 32
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 33
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 34
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 35
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 36
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 37
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 38
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 39
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 40
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 41
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 42
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 43
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 44
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 45
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 46
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 47
2024-03-18 18:35:12 2024/03/18 23:35:12 [notice] 1#1: start worker process 48
```

**Comment by dannongruver** [comment]
I added then environment variable for FORWARDED_ALLOW_IPS to address `--forwarded-allow-ips='*'`. Unfortunately, there's no environment variable for `--proxy-headers`. 

Here's the updated docker-compose yaml
```
version: '3.8'
services:
    ollama:
        deploy:
            resources:
                reservations:
                    devices:
                        - driver: nvidia
                          count: all
                          capabilities:
                              - gpu
        volumes:
            - C:\Projects\LocalLLM\ollama:/root/.ollama
        ports:
            - 11434:11434
        container_name: ollama
        image: ollama/ollama
        command: 
        networks:
          isolated:
    open-webui:
        environment:
            - FORWARDED_ALLOW_IPS=*          
        ports:
            - 8080:8080
        extra_hosts:
            - host.docker.internal:host-gateway
        container_name: open-webui
        restart: always
        image: ghcr.io/open-webui/open-webui:main
        depends_on:
            - ollama
        networks:
            - isolated

    nginx:
        container_name: nginx
        ports:
            - 80:80
        image: nginx:latest
        depends_on:
            - open-webui
        networks:
            - isolated    
            - internet
networks:
    isolated:
        driver: bridge
        internal: true
    internet:
        driver: bridge
```

**Reply by dannongruver** [reply]
Also looking thru the open-webui code [backend/start.sh](https://github.com/open-webui/open-webui/blob/d865b9fe5932af6c6deb7a8503e4a66ce5529d36/backend/start.sh), --forwarded-allow-ips '*' was already applied...
```
WEBUI_SECRET_KEY="$WEBUI_SECRET_KEY" exec uvicorn main:app --host 0.0.0.0 --port "$PORT" --forwarded-allow-ips '*'
```

just missing the --proxy-headers

**Reply by dannongruver** [reply]
According to Uvicorn, `--proxy-headers` defaults to enabled:

>--proxy-headers / --no-proxy-headers - Enable/Disable X-Forwarded-Proto, X-Forwarded-For, X-Forwarded-Port to populate remote address info. Defaults to enabled, but is restricted to only trusting connecting IPs in the forwarded-allow-ips configuration. 
>https://www.uvicorn.org/settings/

**Comment by dannongruver** [comment]
ok...this turned out to be a browser caching issue. refreshed cache and the open-webui displays and is functional. here's nginx minimal config that worked:

```
  server {
    listen 80;
    client_max_body_size 4G;
    server_name localhost;

    location / {     

      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection $connection_upgrade;
      proxy_pass http://open-webui:8080;

    }
        
  }

  map $http_upgrade $connection_upgrade {
    default upgrade;
    '' close;
  }

```

**Reply by Mr-Sunglasses** [reply]
> ok...this turned out to be a browser caching issue. refreshed cache and the open-webui displays and is functional. here's nginx minimal config that worked:
> 
> ```
>   server {
>     listen 80;
>     client_max_body_size 4G;
>     server_name localhost;
> 
>     location / {     
> 
>       proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
>       proxy_set_header X-Forwarded-Proto $scheme;
>       proxy_set_header Upgrade $http_upgrade;
>       proxy_set_header Connection $connection_upgrade;
>       proxy_pass http://open-webui:8080;
> 
>     }
>         
>   }
> 
>   map $http_upgrade $connection_upgrade {
>     default upgrade;
>     '' close;
>   }
> ```

Thanks @dannongruver 😄

**Comment by dannongruver** [comment]
After getting the portal to function, it was having the issue that it couldn't connect to host.docker.internal:11434. This required changing the host's host.docker.internal to a hard-coded internal docker IP.  

Here's the compose yaml to create an internal network w/ specific subnet and assign an IP from that subnet to ollama container: 
```
version: '3.8'
services:
    ollama:
        deploy:
            resources:
                reservations:
                    devices:
                        - driver: nvidia
                          count: all
                          capabilities:
                              - gpu
        ports:
            - 11434:11434
        container_name: ollama
        image: ollama/ollama
        restart: unless-stopped
        networks:
          isolated:
            ipv4_address: 172.18.0.22
    open-webui:
        ports:
            - 8080:8080
        extra_hosts:
            - host.docker.internal:172.18.0.22
        container_name: open-webui
        restart: unless-stopped
        image: ghcr.io/open-webui/open-webui:main
        depends_on:
            - ollama
        networks:
            - isolated

    nginx:
        container_name: nginx
        ports:
            - 80:80
        image: nginx:latest
        depends_on:
            - open-webui
        restart: unless-stopped
        networks:
            - isolated    
            - internet
networks:
    isolated:
        driver: bridge
        internal: true
        ipam:
            driver: default
            config:
                - subnet: "172.18.0.0/16"
    internet:
        driver: bridge
``` 

Finally...both the ollama and open-webui containers are restricted to internal access and open-webui is reverse proxied thru nginx (see nginx config above).

@tjbck: this docker compose file and nginx config may be useful to some others that **_really_** want to lock down local llms.

**Reply by justinh-rahb** [reply]
> After getting the portal to function, it was having the issue that it couldn't connect to host.docker.internal:11434. This required changing the host's host.docker.internal to a hard-coded internal docker IP.

There's no need to change the `host.docker.internal` to a static internal Docker IP. The `host.docker.internal` alias is intended for scenarios where Ollama is installed directly on the host machine, outside of Docker. If Ollama is part of your Docker stack, the correct approach is to communicate with it through the Docker network. This can be done using the container name of Ollama, which Docker’s internal DNS resolves.

**Reply by dannongruver** [reply]
Unfortunately changing the compose file to `host.docker.internal:ollama`, didn’t work or if you mean that it should work without changing `host.docker.internal:host-gateway` then that doesn’t work either for an internal docker network. Not sure why.

**Reply by justinh-rahb** [reply]
> Unfortunately changing the compose file to `host.docker.internal:ollama`, didn’t work or if you mean that it should work without changing `host.docker.internal:host-gateway` then that doesn’t work either for an internal docker network. Not sure why.
> 

You misunderstand, I am saying that you **wouldn't** use that in this scenario, because your Ollama is _in the stack, not outside of it_. Therefore, you should use the container's name as the hostname of the URL, not an IP or the `host.docker.internal` alias.

**Reply by dannongruver** [reply]
I guess I’m not understanding. I also tried commenting out extra_hosts completely but that didn’t work either. Might be best to post an example of what you’re saying. I’m not the best at docker.

**Reply by justinh-rahb** [reply]
```yaml
version: '3.8'
services:
  ollama:
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: all
              capabilities:
                - gpu
    container_name: ollama
    image: ollama/ollama
    restart: unless-stopped
    networks:
      isolated:

  open-webui:
    container_name: open-webui
    restart: unless-stopped
    image: ghcr.io/open-webui/open-webui:main
    environment:
      - 'OLLAMA_BASE_URL=http://ollama:11434'
    depends_on:
      - ollama
    networks:
      - isolated

  nginx:
    container_name: nginx
    ports:
      - "80:80"
    image: nginx:latest
    depends_on:
      - open-webui
    restart: unless-stopped
    networks:
      - isolated
      - internet

networks:
  isolated:
    driver: bridge
    internal: true
  internet:
    driver: bridge
```

---

I've made a few adjustments to your Docker Compose file to enhance the isolation between `ollama` and `open-webui`, while also simplifying the overall configuration. Here are the key changes and the reasons behind them:

1. **Removed External Port Mappings for `ollama` and `open-webui`**: By eliminating these port mappings, these services are no longer directly accessible from outside the Docker environment. This is a crucial step in isolating these services, ensuring that communication between `ollama` and `open-webui` occurs internally within the Docker network. External access is handled through `nginx`, which is the only service exposed to the outside, serving as a controlled entry point.

2. **Removed `extra_hosts` Configuration**: This was unnecessary because Docker’s internal DNS allows services within the same network to resolve each other by service names. This simplification takes advantage of Docker's built-in service discovery mechanisms, making your setup more straightforward and reliable.

3. **Environment Variable for `open-webui`**: I added the `OLLAMA_BASE_URL=http://ollama:11434` environment variable to `open-webui`. This explicitly defines how `open-webui` should connect to `ollama`, using Docker's internal networking. This change ensures that `open-webui` knows where to find `ollama` without relying on external access or hard-coded IPs.

4. **Network Adjustments**: I kept your network setup as is, with `isolated` being an internal network for secure service-to-service communication, and `internet` for controlled external access via `nginx`. This maintains a clear separation between internal and external communications, reinforcing the isolation and security of your services.

These changes aim to simplify your Docker setup, enhance security through isolation, and leverage Docker's built-in features for service discovery and networking. This approach reduces the complexity of your configuration and makes your services more secure and easier to manage.

**Reply by dannongruver** [reply]
Ok, gotcha. I tested and that works perfectly. Also the `OLLAMA_BASE_URL` **`environment`** variable is much nicer than having to create subnet with IPs. Thank you for this!
```
Side note: For anyone using docker on windows you'll might need to remove the single quotes like this: 
`- OLLAMA_BASE_URL=http://ollama:11434`
```

**Reply by justinh-rahb** [reply]
EDIT: I've just realized that your previous example compose that I modified is lacking the volume mounts for Ollama and WebUI to persist their data, you're going to lose this as soon as you `compose down` this stack. I'd highly recommend sticking to the default compose files we provide, and using the `run-compose.sh` script.

To be straightforward, if your aim is to restrict external access to the `ollama` container, there was no need to expose its port (11434) outside of your Docker stack’s network in the first place. The creation of separate `internet` and `isolated` networks as you've done doesn’t add significant value, except for limiting `ollama` and `open-webui` to only communicate within their isolated network, which could negatively impact their functionality. This approach might only make sense if you have concerns about the security of these software components. However, if that were the case, one might question the decision to use them at all. As for the inclusion of an `nginx` proxy in front of `open-webui`, it doesn’t provide substantial advantages unless you’re implementing HTTPS, which isn't the case here.

**Reply by dannongruver** [reply]
Yeah you’re right, the goal is to block self hosted apps from calling home. Trust but verify 🤷. The ip/subnet solution works for any docker that doesn’t require online access. The `environment` variable is specifically better for open-webui. 

I think this compose yaml makes the strongest statement for privacy: open-webui will never share your data. It can’t. Your help here shows your dedication to that goal. Thank you!

Finally…yeah my comments redacted all the `volume` sections for all 3 containers - I’ve restarted multiple times during testing and didn’t lose anything. Thanks for looking out for that.

Thank you for the work on open-webui!
